cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java

```diff
@@ -443,4 +500,147 @@ private long findProgramFinishTime(List<Message> messages) {
+      // One time set up for decoding.
+      this.cachemap = new HashMap<>();
+      decodingOutputStream = new MessagingMetricsProcessorService.PayloadInputStream();
+      decoder = new StringCachingDecoder(new BinaryDecoder(decodingOutputStream), cachemap);
+      this.encodeOutputStream = new ByteArrayOutputStream(1024);
+      encoder = new BinaryEncoder(encodeOutputStream);
+      this.reader = reader;
+      this.schema = schema;
+      this.recordWriter = recordWriter;
+    }
+
+    @Override
+    public Iterator<Message> aggregate(Iterator<Message> input) {
+      if (!input.hasNext()) {
+        return input;
+      }
+      ArrayList<Message> messages = new ArrayList<>();
+      while (input.hasNext()) {
+        Message message = input.next();
+        messages.add(message);
+      }
+
+      ArrayList<MetricValues> decoded = new ArrayList<>();
+
+      // Decode the messages into MetricValues.
+      String lastMessageId = null;
+      for (Message message : messages) {
+        lastMessageId = message.getId();
+        decodingOutputStream.reset(message.getPayload());
+        try {
+          decoded.add(reader.read(decoder, schema));
+        } catch (IOException e) {
+          LOG.warn("Failed to decode metric message.", e);
+        }
+      }
+      cachemap.clear();
+
+      // Aggregate the metrics.
+      ArrayList<MetricValues> aggregated = aggregateMetricValues(decoded);
+      ArrayList<Message> result = new ArrayList<>();
+
+      // Encode metrics back to messages.
+      for (MetricValues metricValues : aggregated) {
+        // Encode the metric value into a message.
+        encodeOutputStream.reset();
+        try {
+          recordWriter.encode(metricValues, encoder);
+          byte[] payload = encodeOutputStream.toByteArray();
+          // creating a final copy to access in inner class.
+          final String finalLastMessageId = lastMessageId;
+          result.add(new Message() {
+            @Override
+            public String getId() {
+              // TopicRelayers keep track of the last message ID
+              // to fetch messages in the next iteration.
+              // Message IDs are generated by the messaging service. Return the last message ID before aggregation
+              // every time.
+              return finalLastMessageId;
+            }
+
+            @Override
+            public byte[] getPayload() {
+              return payload;
+            }
+          });
+        } catch (IOException e) {
+          LOG.warn("Failed to encode metric to message payload.", e);
+        }
+      }
+
+      return result.iterator();
+    }
+
+    private ArrayList<MetricValues> aggregateMetricValues(ArrayList<MetricValues> metrics) {
+      ArrayList<MetricValues> result = new ArrayList<>();
+
+      // Group metrics with the same tags.
+      HashMap<Map<String, String>, ArrayList<MetricValues>> grouped = new HashMap<>();
```

Comment:
Use `Map` as the type.

Reply:
Also use `List` in the parameterized type.

Reply:
Noted.